core/src/banking_simulation.rs

```diff
@@ -0,0 +1,908 @@
+                ChannelLabel::Dummy => unreachable!(),
+            };
+            sender.send(batches_with_stats).unwrap();
+
+            logger.on_sending_batches(&simulation_duration, label, batch_count, tx_count);
+            if self.exit.load(Ordering::Relaxed) {
+                break;
+            }
+        }
+        logger.on_terminating();
+        drop(self.timed_batches_to_send);
+        // hold these senders in join_handle to control banking stage termination!
+        (
+            self.non_vote_sender,
+            self.tpu_vote_sender,
+            self.gossip_vote_sender,
+        )
+    }
+}
+
+struct SimulatorLoop {
+    bank: BankWithScheduler,
+    parent_slot: Slot,
+    first_simulated_slot: Slot,
+    freeze_time_by_slot: FreezeTimeBySlot,
+    base_event_time: SystemTime,
+    poh_recorder: Arc<RwLock<PohRecorder>>,
+    simulated_leader: Pubkey,
+    bank_forks: Arc<RwLock<BankForks>>,
+    blockstore: Arc<Blockstore>,
+    leader_schedule_cache: Arc<LeaderScheduleCache>,
+    retransmit_slots_sender: Sender<Slot>,
+    retracer: Arc<BankingTracer>,
+}
+
+impl SimulatorLoop {
+    fn enter(
+        self,
+        base_simulation_time: SystemTime,
+        sender_thread: EventSenderThread,
+    ) -> (EventSenderThread, Sender<Slot>) {
+        sleep(WARMUP_DURATION);
+        info!("warmup done!");
+        self.start(base_simulation_time, sender_thread)
+    }
+
+    fn start(
+        self,
+        base_simulation_time: SystemTime,
+        sender_thread: EventSenderThread,
+    ) -> (EventSenderThread, Sender<Slot>) {
+        let logger = SimulatorLoopLogger {
+            simulated_leader: self.simulated_leader,
+            base_event_time: self.base_event_time,
+            base_simulation_time,
+            freeze_time_by_slot: self.freeze_time_by_slot,
+        };
+        let mut bank = self.bank;
+        loop {
+            if self.poh_recorder.read().unwrap().bank().is_none() {
+                let next_leader_slot = self.leader_schedule_cache.next_leader_slot(
+                    &self.simulated_leader,
+                    bank.slot(),
+                    &bank,
+                    Some(&self.blockstore),
+                    GRACE_TICKS_FACTOR * MAX_GRACE_SLOTS,
+                );
+                debug!("{next_leader_slot:?}");
+                self.poh_recorder
+                    .write()
+                    .unwrap()
+                    .reset(bank.clone_without_scheduler(), next_leader_slot);
+                info!("Bank::new_from_parent()!");
+
+                logger.log_jitter(&bank);
+                bank.freeze();
+                let new_slot = if bank.slot() == self.parent_slot {
+                    info!("initial leader block!");
+                    self.first_simulated_slot
+                } else {
+                    info!("next leader block!");
+                    bank.slot() + 1
+                };
+                info!("new leader bank slot: {new_slot}");
```

Comment:
we might not be leader anymore at this point, right? this line seems confusing.

Reply:
done: c83a7fed205bd9fc92871fb37b6758442beca3af